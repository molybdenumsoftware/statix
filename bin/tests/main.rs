mod util {
    #[macro_export]
    macro_rules! test_lint {
        ($tname:ident, $($tail:tt)*) => {
            test_lint!($tname);
            test_lint!($($tail)*);
        };
        ($tname:ident, $($tail:tt)*) => {
                test_lint!($tname);
                test_lint!($($tail)*);
        };
        ($tname:ident) => {
            paste::paste! {
                #[test]
                fn [<$tname _lint>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_lint"), file_path, &["check"]);
                }

                #[test]
                fn [<$tname _fix>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_fix"), file_path, &["fix", "--dry-run"]);
                }
            }

        };
    }
}

fn test_cli(test_name: &str, file_path: &str, args: &[&str]) {
    let output = std::process::Command::new("cargo")
        .arg("run")
        .arg("--")
        .args(args)
        .arg(file_path)
        .output()
        .expect("command runs successfully");

    let stdout = strip_ansi_escapes::strip(output.stdout).unwrap();
    let stdout = String::from_utf8(stdout).unwrap();

    insta::assert_snapshot!(test_name, &stdout);
}

test_lint! {
    collapsible_let_in,
    eta_reduction,
    useless_parens,
    empty_pattern,
    redundant_pattern_bind,
    unquoted_uri,
    empty_inherit,
    deprecated_to_path,
    useless_has_attr,
    repeated_keys,
    empty_list_concat
}
