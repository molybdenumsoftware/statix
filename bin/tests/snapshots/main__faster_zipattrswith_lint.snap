---
source: bin/tests/main.rs
expression: "& out"
---
[W16] Warning: Found lib.zipAttrsWith
   ╭─[data/faster_zipattrswith.nix:3:7]
   │
 3 │   _ = lib.zipAttrsWith (name: values: values) [{ a = 1; } { a = 2; b = 3; }];
   ·       ────────┬───────  
   ·               ╰───────── Prefer builtins.zipAttrsWith over lib.zipAttrsWith
───╯
[W16] Warning: Found lib.zipAttrsWith
   ╭─[data/faster_zipattrswith.nix:6:7]
   │
 6 │   _ = nixpkgs.lib.zipAttrsWith (name: values: values) [{ a = 1; } { a = 2; b = 3; }];
   ·       ────────────┬───────────  
   ·                   ╰───────────── Prefer builtins.zipAttrsWith over nixpkgs.lib.zipAttrsWith
───╯
